Refactor verification options

richsalz opened the pull request:
This refactors all verification options to common documentation in openssl.pod
I also rewrote the prose text about how verification works.

I'd like #10118 to get reviewed and merged first because the options that control the trust chain (no-cafile, etc) affect this and I can add the necessary ```L<>``` link.

richsalz commented:
I did some editing on the verification options in openssl.pod, and I tried to put them into a more sensible ordering.  Once this is reviewed and merged, I will change ```OPT_V_OPTIONS```
in include/opt.h to match the ordering.

richsalz commented:
This is now based on #10159 and needs that to go in first.

richsalz commented:
Rebased since #10159 was merged.

richsalz commented:
re-pushing to restart hung CI jobs.

richsalz commented:
rebased and fixed conflicts.

vdukhovni commented on doc/man1/openssl-s_client.pod.in:
```diff
@@ -40,36 +40,9 @@ B<openssl> B<s_client>
-[B<-policy> I<arg>]
-[B<-policy_check>]
-[B<-policy_print>]
 [B<-purpose> I<purpose>]
```
Did you mean to leave "-purpose" here?

richsalz replied:
No, good catch.  Removed.

vdukhovni commented on doc/man1/openssl-verify.pod.in:
```diff
@@ -326,63 +123,6 @@ Certificates must be in PEM format.
-by this command: wherever possible an attempt is made to
-continue after an error whereas normally the verify operation would halt on
-the first error. This allows all the problems with a certificate chain to be
-determined.
```
Is the above paragraph about the verification mode in verify(1) preserved somewhere?

richsalz replied:
Yes, as described in the openssl command, okay?

vdukhovni commented on doc/man1/openssl.pod:
```diff
@@ -778,11 +778,266 @@ client.
+The input format for the extra certifcate and key, respectively.
+See L<openssl(1)/Format Options> for details.
+<<<<<<< dc5af7967d7a4b6afd777a460f0c8918768e256e
+=======
```
Please address the merge conflict.

richsalz replied:
aargh, fixing all of them.  thanks.

vdukhovni commented on doc/man1/openssl.pod:
```diff
@@ -778,11 +778,266 @@ client.
+Specify whether the application should build the certificate chain to be
+provided to the server for the extra certificates via the B<-xkey>,
+B<-xcert>, and B<-xchain> options.
+>>>>>>> Move verify options to openssl.pod
```
DItto.

richsalz replied:
fixed.

vdukhovni commented on doc/man1/openssl.pod:
```diff
@@ -778,11 +778,266 @@ client.
+
+Many OpenSSL commands verify certificates.  The commands try to proceed
+after any validation error, in order to allow all the problems with a
+certificate chain to be determined.
```
Is that true for commands other than `verify(1)`?

richsalz replied:
They all *try* to proceed.  Sometimes they don't.  So, yes.  Weasel-wording :)

vdukhovni replied:
The reason I asked, is that `s_client(1)` and `s_server(1)` have a `-verify_return_error` option, that makes the first error fatal, and the command then returns a non-zero exit code when peer chain verification fails.  They also use a different callback function than the one used by `verify(1)`.  The latter continues only for some errors and not others:
```
        /*
         * Pretend that some errors are ok, so they don't stop further
         * processing of the certificate chain.  Setting ok = 1 does this.
         * After X509_verify_cert() is done, we verify that there were
         * no actual errors, even if the returned value was positive.
         */    
        switch (cert_error) {
        case X509_V_ERR_NO_EXPLICIT_POLICY:
            policies_print(ctx);
            /* fall thru */                                                 
        case X509_V_ERR_CERT_HAS_EXPIRED:
            /* Continue even if the leaf is a self signed cert */
        case X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT:
            /* Continue after extension errors too */
        case X509_V_ERR_INVALID_CA:
        case X509_V_ERR_INVALID_NON_CA:
        case X509_V_ERR_PATH_LENGTH_EXCEEDED:
        case X509_V_ERR_INVALID_PURPOSE:
        case X509_V_ERR_CRL_HAS_EXPIRED:
        case X509_V_ERR_CRL_NOT_YET_VALID:
        case X509_V_ERR_UNHANDLED_CRITICAL_EXTENSION:
            ok = 1;
        }
```
While the callback used by `s_server` and `s_client` continues on almost all errors:
```
    if (!ok) {
        BIO_printf(bio_err, "verify error:num=%d:%s\n", err,
                   X509_verify_cert_error_string(err));
        if (verify_args.depth < 0 || verify_args.depth >= depth) {
            if (!verify_args.return_error)
                ok = 1;
            verify_args.error = err;
        } else {
            ok = 0;
            verify_args.error = X509_V_ERR_CERT_CHAIN_TOO_LONG;
        }
    }
```
except for its own synthetic "chain too long" error, which is a mistake, since I fixed the handling of long chains in libcrypto as of 1.1.0. The special case should simply be removed. This will allow chains one longer than the depth limit argument, since the documented limit bounds the number of intermediate CAs, not the index of the last element.

vdukhovni replied:
Actually the other apps also have their own custom callbacks:
```text
apps/cms.c:        X509_STORE_set_verify_cb(store, cms_cb);
apps/smime.c:        X509_STORE_set_verify_cb(store, smime_cb);
apps/ts.c:    X509_STORE_set_verify_cb(cert_ctx, verify_cb);
apps/verify.c:    X509_STORE_set_verify_cb(store, cb);
```
In `cms(1)`, `smime(1)` and `ts(1)` the first error is fatal.

richsalz replied:
Thanks for the detailed comments (and research).  I'll address them over the weekend.

richsalz replied:
I fixed the documentation of the commands you noted above to properly explain the behavior on verification error.

vdukhovni commented on doc/man1/openssl.pod:
```diff
@@ -778,11 +778,266 @@ client.
+is found the remaining lookups are from the trusted certificates. The root CA
+is always looked up in the trusted certificate list: if the certificate to
+verify is a root certificate then an exact match must be found in the trusted
+list.
```
Here, you'll need to reconcile with #10587 or perhaps 10587 refactored to follow your patch...

richsalz replied:
Yes, that's very very frustrating.  This PR has sat her since *OCTOBER 9* and nobody has looked at it until now.  I have rebased, posted comments, and so on.

If #10587 goes in before this, then I hope someone else will provide a patch because I doubt that I will do it. There is a cost to ignoring PR's and this is a prime example. By letting this languish, the project has forced someone (@ddvo or me), who is only trying to help improve documentation and code, to do needless extra work. Maybe the project will try to be more diligent in the future so that this type of problem can be avoided.

vdukhovni replied:
I think your PR should be merged first, and then #10587. I hope that's OK.
Not sure how important the fine distinctions between s_client/s_server and the other apps (verify, ...) are to document, but lean towards documenting the actual behaviour.

richsalz commented:
Thanks for reading this, @vdukhovni I appreciate it.  Fixup commit pushed.

richsalz commented:
I had to rebase because of a conflict.  I fixed various failure descriptions and moved the verification error description all to the man3 file.  That file now has the define name and the text.  Should the text, which is basically a repetition of the text description, be removed? I think so.

richsalz commented:
Travis failure is a timeout.

mspncp commented:
I restarted the job.

vdukhovni commented on doc/man1/openssl-cms.pod.in:
```diff
@@ -462,16 +440,9 @@ portion of a message so they may be included manually. If signing
 
-Set various certificate chain validation options. See the
-L<openssl-verify(1)> manual page for details.
+Any verification errors cause the command to exit.
```
Perhaps not a show-stopper, but I think better would be:
```
Certificate verification terminates with a failure as soon as the first error detected. Additional errors if any are not reported.
```

vdukhovni replied:
Should something similar be said in the `ocsp(1)` manpage?

richsalz replied:
I agree that the wording should be the same across the manpages.  I prefer mine because it says that the *program* exits, while your suggestion says *verification terminates* without saying the effect on the program.

vdukhovni replied:
But the program does not exit at that point, it merely returns verification failure to the caller, with the caller doing whatever it does when verification fails (which may be exit, but I did not check). If you're sure that the effect is effectively an immediate exit, I'll grudgingly accept the less precise description. :-)

richsalz replied:
Yes, verify will exit if it's check() routine fails, CMS exits on verification failure, OCSP same, etc.  Since the "caller" is the application program, and the manpage is describing the application program, I think it's correct.

vdukhovni commented on doc/man1/openssl-s_client.pod.in:
```diff
@@ -712,6 +675,11 @@ I<localhost> on port I<4433>.
 
+{- $OpenSSL::safe::opt_v_item -}
+
+Verification errors are display, for debugging, but the command will
```
Typo: s/display/displayed/

richsalz replied:
fixed, thanks.

vdukhovni commented on doc/man1/openssl-s_server.pod.in:
```diff
@@ -727,6 +682,11 @@ by the client in binary mode.
 
+{- $OpenSSL::safe::opt_v_item -}
+
+Verification errors are display, for debugging, but the command will
```
Ditto, but could mention that this applies only when client certificates are solicited by server and provided by the client.

richsalz replied:
Put "If the server requests a client certificate, than ..." at the start of the sentence.  Fixed the typo.

vdukhovni commented on doc/man1/openssl-smime.pod.in:
```diff
@@ -283,16 +262,9 @@ portion of a message so they may be included manually. If signing
 
-Set various options of certificate chain verification. See
-L<openssl-verify(1)> manual page for details.
+Any verification errors cause the command to exit.
```
See comment re `cms(1)`.

vdukhovni commented on doc/man1/openssl-ts.pod.in:
```diff
@@ -363,17 +331,9 @@ certificate. This file must contain the TSA signing certificate and
-verification.  See L<openssl-verify(1)>.
+{- $OpenSSL::safe::opt_v_item -}
+
+Any verification errors cause the command to exit.
```
See `cms(1)`.

vdukhovni commented on doc/man1/openssl-verify.pod.in:
```diff
@@ -314,6 +109,23 @@ with a B<-->.
 
+{- $OpenSSL::safe::opt_v_item -}
+
+This command ignores many errors, in order to allow all the problems with a
```
It isn't that errors are "ignored", rather verification continues after reporting the error, the final outcome is still verification failure if any of these errors are encountered. Frankly, we should probably continue for all errors, not just the special-cased ones.  I can't think of a good reason to immediately abort verification in the other error cases...

vdukhovni replied:
The end result is that multiple errors are reported if present, but only up to the first error that is not one of the special ones. Again, not sure why...

richsalz replied:
I think "ignores errors" is correct since the function sets "ok=1"  Is that wrong?

I agree the behavior is kinda half-way and should be fixed in another PR.  I'm just trying to improve the documentation in this PR.

vdukhovni replied:
Setting `ok = 1` just means **continue**, it does not clear the error. Ultimately `sctx->error` is still set.

richsalz replied:
So maybe just: "If a certificate chain has multiple problems, this program tries to display all of them." And remove the specific list of errors?

vdukhovni replied:
Perhaps so. And I'll open an issue about whether it should continue on all errors (if I don't forget, and nobody else beats me to it).

richsalz replied:
Beat you to it: #10750 :)

richsalz commented:
Ping for a second review.  Should be easy, it made it past one of the harder reviewers already :)  And #10587 is kinda waiting on this.

mspncp commented on doc/man1/openssl-smime.pod.in:
```diff
@@ -19,33 +19,10 @@ B<openssl> B<smime>
 [B<-crlfeol>]
 [B<-I<cipher>>]
 [B<-in> I<file>]
-[B<-attime> I<timestamp>]
```
Hmmm... I find it hard to verify that no option got lost on the way. Any hints how to achieve this?

richsalz replied:
I guess look at the "opt_v_synopsis" variable in one window, and compare it against the "-" lines in another?

mspncp replied:
It occurred to me that I could also compare the generated files. But I didn't find the time to do it yet.

richsalz commented on crypto/x509/x509_txt.c:
```diff
@@ -133,7 +133,7 @@ const char *X509_verify_cert_error_string(long n)
         return "CRL path validation error";
     case X509_V_ERR_PATH_LOOP:
-        return "Path Loop";
+        return "path Loop";
```
fixup commit pushed.  good catch!

richsalz commented:
There were conflicts, so I did a squash/rebase.  The conflicts were simple, but this probably needs another review from @vdukhovni (as well as a committer).

mspncp commented:
I generated the pod files using 8062c027e9cdef644f1e63b210384e760c65a354 and its parent, see attached [comparison.zip](https://github.com/openssl/openssl/files/4026809/comparison.zip). For better comparison I sorted the options in the SYNOPSIS section.

The diff shows some duplicates (`-CAfile`, `-CApath`, `-certfile`) and some options which have disappeared (`-no_check_time`) or appeared (`-verbose`). I think the duplicates need to be removed and the other changes need a closer look and/or an explanation why they were changed.

See for example openssl-cms.pod:

```diff
msp@msppc:~/src/openssl$ diff before/openssl-cms.pod after/openssl-cms.pod
--- before/openssl-cms.pod	2020-01-06 16:25:32.978669176 +0100
+++ after/openssl-cms.pod	2020-01-06 16:26:16.407071393 +0100
@@ -15,6 +15,8 @@
 
 B<openssl> B<cms>
 [B<-CAfile> I<file>]
+[B<-CAfile> I<file>]
+[B<-CApath> I<dir>]
 [B<-CApath> I<dir>]
 [B<-CAstore> I<uri>]
 [B<-EncryptedData_encrypt>]
@@ -25,6 +27,7 @@
 [B<-binary>]
 [B<-cades>]
 [B<-certfile> I<file>]
+[B<-certfile> I<file>]
 [B<-certsout> I<file>]
 [B<-check_ss_sig>]
 [B<-cmsout>]
@@ -56,10 +59,11 @@
 [B<-keyopt> I<name>:I<parameter>]
 [B<-md> I<digest>]
 [B<-no-CAfile>]
+[B<-no-CAfile>]
+[B<-no-CApath>]
 [B<-no-CApath>]
 [B<-no-CAstore>]
 [B<-no_alt_chains>]
-[B<-no_check_time>]
 [B<-noattr>]
 [B<-nocerts>]
 [B<-nodetach>]
@@ -102,6 +106,7 @@
 [B<-trusted_first>]
 [B<-uncompress>]
 [B<-use_deltas>]
+[B<-verbose>]
 [B<-verify>]
 [B<-verify_depth> I<num>]
 [B<-verify_email> I<email>]
@@ -476,14 +481,16 @@
 
 =item B<-attime>, B<-check_ss_sig>, B<-crl_check>, B<-crl_check_all>,
 B<-explicit_policy>, B<-extended_crl>, B<-ignore_critical>, B<-inhibit_any>,
-B<-inhibit_map>, B<-no_alt_chains>, B<-no_check_time>, B<-partial_chain>, B<-policy>,
+B<-inhibit_map>, B<-no_alt_chains>, B<-partial_chain>, B<-policy>,
 B<-policy_check>, B<-policy_print>, B<-purpose>, B<-suiteB_128>,
 B<-suiteB_128_only>, B<-suiteB_192>, B<-trusted_first>, B<-use_deltas>,
 B<-auth_level>, B<-verify_depth>, B<-verify_email>, B<-verify_hostname>,
 B<-verify_ip>, B<-verify_name>, B<-x509_strict>
 
-Set various certificate chain validation options. See the
-L<openssl-verify(1)> manual page for details.
+Set various options of certificate chain verification.
+See L<openssl(1)/Verification Options> for details.
+
+Any verification errors cause the command to exit.
 
 =item B<-CAfile> I<file>, B<-no-CAfile>, B<-CApath> I<dir>, B<-no-CApath>,
 B<-CAstore> I<uri>, B<-no-CAstore>
```
_Note: I did not go through all the diffs yet._

richsalz commented:
The duplicates (CApath,no-CApath,CAfile,no-CAfile,certfile) where rebase mistakes that slipped through; fixup commit pushed.  The no_check_time isn't implemented so I removed it.  As you go through the review, I'll fix each one in turn.  Thanks!

mspncp commented:
> The no_check_time isn't implemented so I removed it.

Are you sure it's not implemented?

**$ git grep -in no_check_time apps**
[apps/include/opt.h:30:](https://github.com/openssl/openssl/blob/8062c027e9/apps/include/opt.h#L30)         OPT_V_PARTIAL_CHAIN, OPT_V_NO_ALT_CHAINS, OPT_V_NO_CHECK_TIME, \
[apps/include/opt.h:82:](https://github.com/openssl/openssl/blob/8062c027e9/apps/include/opt.h#L82)         { "no_check_time", OPT_V_NO_CHECK_TIME, '-', "ignore certificate validity time" }, \
[apps/include/opt.h:115:](https://github.com/openssl/openssl/blob/8062c027e9/apps/include/opt.h#L115)         case OPT_V_NO_CHECK_TIME:  \
[apps/lib/opt.c:608:](https://github.com/openssl/openssl/blob/8062c027e9/apps/lib/opt.c#L608)     case OPT_V_NO_CHECK_TIME: 
[apps/lib/opt.c:609:](https://github.com/openssl/openssl/blob/8062c027e9/apps/lib/opt.c#L609)         X509_VERIFY_PARAM_set_flags(vpm, X509_V_FLAG_NO_CHECK_TIME);

richsalz commented:
Thanks @mspncp you're right.  Lucky a simple fix in only one file fixes it everywhere :)

mspncp commented:
**-no_check_time**

Did you restore -no_check_time already?

**More duplicates**

I see more duplicates of the following options:

```
-CAfile, -CApath, -no-CAfile, -no-CApath
```

I'm sure you'll find them all using grep ;-)


**Verbosity**

The -verbose option was somehow 'inverted': it is now documented where it previously wasn't and vice versa.

```
msp@msppc:~/src/openssl$ grep -- -verbose before/*
before/openssl-verify.pod:[B<-verbose>]
before/openssl-verify.pod:=item B<-verbose>

msp@msppc:~/src/openssl$ grep -- -verbose after/*
after/openssl-cms.pod:[B<-verbose>]
after/openssl-ocsp.pod:[B<-verbose>]
after/openssl-s_client.pod:[B<-verbose>]
after/openssl-s_server.pod:[B<-verbose>]
after/openssl-smime.pod:[B<-verbose>]
after/openssl-ts.pod:[B<-verbose>]
```

According to my quick test, none of the latter commands supports this flag:

```
msp@msppc:~/src/openssl$ for cmd in cms ocsp s_client smime ts; do openssl $cmd -verbose ; done

cms: Unrecognized flag verbose
ocsp: Unrecognized flag verbose
ocsp: Use -help for summary.
s_client: Option unknown option -verbose
s_client: Use -help for summary.
smime: Unrecognized flag verbose
smime: Use -help for summary.
ts: Unrecognized flag verbose
ts: Use -help for summary.
```

mspncp commented:
Apropos: the missing-flag error messages are also very higgledy-piggledy:

```
cms: Unrecognized flag verbose
ocsp: Unrecognized flag verbose
ocsp: Use -help for summary.
s_client: Option unknown option -verbose
s_client: Use -help for summary.
smime: Unrecognized flag verbose
smime: Use -help for summary.
ts: Unrecognized flag verbose
ts: Use -help for summary.
```

richsalz commented:
Could you open a separate issue for https://github.com/openssl/openssl/pull/10132#issuecomment-571358064 ?  Thanks.

mspncp commented:
> Could you open a separate issue for [#10132 (comment)](https://github.com/openssl/openssl/pull/10132#issuecomment-571358064) ? Thanks.

Done, see #10773.

richsalz commented:
> **-no_check_time**
> 
> Did you restore -no_check_time already?

Yes but I didn't push waiting for your other feedback.

> **More duplicates**
> ...
> -CAfile, -CApath, -no-CAfile, -no-CApath

I found them in ocsp, s_time, smime, ts, and verify manpages.

> I'm sure you'll find them all using grep ;-)

Hope so :)
 
> **Verbosity**

I removed the -verbose flag from the synopsis (which is the only place it appeared) from cms, ocsp, s_client, s_server, smime, and ts manpages.

Thanks for your attention to detail!  BTW, you might find "find-doc-nits -c" interesting.  And issue #9730

richsalz commented:
Rebased to address conflicts after #10128 was merged.

t8m commented:
The Travis failure is relevant.

richsalz commented:
Fixed; typo was in doc/perlvars.pm

mspncp commented:
## Lost and Found

I feel a little bit dizzy after chasing all those options which have moved around and checking whether no ~sheep~ options got lost. In particular those options which were moved into openssl.pod are hard to track. (Which doesn't mean at all that I dislike your idea to collect these common options in a common place. On the contrary, I think it's a very good thing to do.)

The following diff shows a comparison of the sorted options of the respective SYNOPSIS section. Please be so kind and check the diffs once more. Some changes are ok, because only the name of the option arg has changed. But there are still options which disappeared or appeared.

```diff
diff -u -r before/openssl-s_client.pod after/openssl-s_client.pod
--- before/openssl-s_client.pod	2020-01-10 23:35:37.034484373 +0100
+++ after/openssl-s_client.pod	2020-01-10 23:35:51.687642628 +0100
@@ -18,6 +18,7 @@
 [B<-build_chain>]
 [B<-cert> I<filename>]
 [B<-cert_chain> I<filename>]
+[B<-certfile> I<file>]
 [B<-certform> B<DER>|B<PEM>]
 [B<-chainCAfile> I<filename>]
 [B<-chainCApath> I<directory>]
diff -u -r before/openssl-s_server.pod after/openssl-s_server.pod
--- before/openssl-s_server.pod	2020-01-10 23:36:23.671988125 +0100
+++ after/openssl-s_server.pod	2020-01-10 23:36:35.399114803 +0100
@@ -12,18 +12,18 @@
 [B<-WWW>]
 [B<-accept> I<val>]
 [B<-allow_no_dhe_kex>]
-[B<-allow_proxy_certs>]
 [B<-alpn> I<val>]
 [B<-anti_replay>]
 [B<-async>]
-[B<-attime> I<intmax>]
-[B<-auth_level> I<int>]
+[B<-attime> I<timestamp>]
+[B<-auth_level> I<num>]
 [B<-brief>]
 [B<-bugs>]
 [B<-build_chain>]
 [B<-cert2> I<infile>]
 [B<-cert> I<infile>]
 [B<-cert_chain> I<infile>]
+[B<-certfile> I<file>]
 [B<-certform> B<DER>|B<PEM>]
 [B<-chainCAfile> I<infile>]
 [B<-chainCApath> I<dir>]
@@ -52,7 +52,7 @@
 [B<-dtls1_2>]
 [B<-dtls>]
 [B<-early_data>]
-[B<-engine> I<val>]
+[B<-engine> I<id>]
 [B<-explicit_policy>]
 [B<-ext_cache>]
 [B<-extended_crl>]
@@ -64,7 +64,6 @@
 [B<-ignore_critical>]
 [B<-inhibit_any>]
 [B<-inhibit_map>]
-[B<-issuer_checks>]
 [B<-key2> I<infile>]
 [B<-key> I<infile>]
 [B<-keyform> B<DER>|B<PEM>]
@@ -110,7 +109,7 @@
 [B<-num_tickets>]
 [B<-partial_chain>]
 [B<-pass> I<val>]
-[B<-policy> I<val>]
+[B<-policy> I<arg>]
 [B<-policy_check>]
 [B<-policy_print>]
 [B<-port> I<+int>]
@@ -119,7 +118,7 @@
 [B<-psk_hint> I<val>]
 [B<-psk_identity> I<val>]
 [B<-psk_session> I<file>]
-[B<-purpose> I<val>]
+[B<-purpose> I<purpose>]
 [B<-quiet>]
 [B<-rand> I<files>]
 [B<-read_buf> I<+int>]
@@ -165,11 +164,11 @@
 [B<-verifyCAfile> I<infile>]
 [B<-verifyCApath> I<dir>]
 [B<-verifyCAstore> I<uri>]
-[B<-verify_depth> I<int>]
-[B<-verify_email> I<val>]
-[B<-verify_hostname> I<val>]
-[B<-verify_ip>]
-[B<-verify_name> I<val>]
+[B<-verify_depth> I<num>]
+[B<-verify_email> I<email>]
+[B<-verify_hostname> I<hostname>]
+[B<-verify_ip> I<ip>]
+[B<-verify_name> I<name>]
 [B<-verify_quiet>]
 [B<-verify_return_error>]
 [B<-writerand> I<file>]
diff -u -r before/openssl-ts.pod after/openssl-ts.pod
--- before/openssl-ts.pod	2020-01-10 23:37:03.668499063 +0100
+++ after/openssl-ts.pod	2020-01-10 23:37:17.217645433 +0100
@@ -35,11 +35,12 @@
 
 B<openssl> B<ts>
 B<-verify>
-[B<-CAfile> I<trusted_certs.pem>]
-[B<-CApath> I<trusted_cert_path>]
-[B<-CAstore> I<trusted_certs_uri>]
+[B<-CAfile> I<file>]
+[B<-CApath> I<dir>]
+[B<-CAstore> I<uri>]
 [B<-attime> I<timestamp>]
 [B<-auth_level> I<num>]
+[B<-certfile> I<file>]
 [B<-check_ss_sig>]
 [B<-crl_check>]
 [B<-crl_check_all>]
@@ -51,7 +52,9 @@
 [B<-in> I<response.tsr>]
 [B<-inhibit_any>]
 [B<-inhibit_map>]
-[B<-issuer_checks>]
+[B<-no-CAfile>]
+[B<-no-CApath>]
+[B<-no-CAstore>]
 [B<-no_alt_chains>]
 [B<-no_check_time>]
 [B<-partial_chain>]
@@ -65,7 +68,6 @@
 [B<-suiteB_192>]
 [B<-token_in>]
 [B<-trusted_first>]
-[B<-untrusted> I<cert_file.pem>]
 [B<-use_deltas>]
 [B<-verify_depth> I<num>]
 [B<-verify_email> I<email>]
@@ -73,6 +75,5 @@
 [B<-verify_ip> I<ip>]
 [B<-verify_name> I<name>]
 [B<-x509_strict>]
-[I<verify options>]
 
 =for openssl ifdef engine
diff -u -r before/openssl-verify.pod after/openssl-verify.pod
--- before/openssl-verify.pod	2020-01-10 23:37:50.699923966 +0100
+++ after/openssl-verify.pod	2020-01-10 23:38:05.519084050 +0100
@@ -6,9 +6,9 @@
 [B<-CApath> I<dir>]
 [B<-CAstore> I<uri>]
 [B<-CRLfile> I<file>]
-[B<-allow_proxy_certs>]
 [B<-attime> I<timestamp>]
-[B<-auth_level> I<level>]
+[B<-auth_level> I<num>]
+[B<-certfile> I<file>]
 [B<-check_ss_sig>]
 [B<-crl_check>]
 [B<-crl_check_all>]
@@ -31,17 +31,11 @@
 [B<-policy_check>]
 [B<-policy_print>]
 [B<-purpose> I<purpose>]
-[B<-show_chain>]
-[B<-sm2-hex-id> I<hex-string>]
-[B<-sm2-id> I<string>]
 [B<-suiteB_128>]
 [B<-suiteB_128_only>]
 [B<-suiteB_192>]
-[B<-trusted> I<file>]
 [B<-trusted_first>]
-[B<-untrusted> I<file>]
 [B<-use_deltas>]
-[B<-verbose>]
 [B<-verify_depth> I<num>]
 [B<-verify_email> I<email>]
 [B<-verify_hostname> I<hostname>]
```


## OPTION subsection order

The order of the OPTION subsections in openssl.pod is debatable. For example it is suprising that the 'Extended Verification Options' precede the 'Verification Options'.


Please take a closer look:

```
doc/man1/openssl.pod:496:=head1 OPTIONS
doc/man1/openssl.pod:501:=head2 Common Options
doc/man1/openssl.pod:519:=head2 Format Options
doc/man1/openssl.pod:622:=head2 Pass Phrase Options
doc/man1/openssl.pod:668:=head2 Trusted Certificate Options
doc/man1/openssl.pod:722:=head2 Random State Options
doc/man1/openssl.pod:752:=head2 Extended Verification Options
doc/man1/openssl.pod:797:=head2 Verification Options
doc/man1/openssl.pod:1037:=head2 Name Format Options
doc/man1/openssl.pod:1187:=head2 Engine Options
```

mspncp commented:
> [Ping for a second review. Should be easy, it made it past one of the harder reviewers already :)](https://github.com/openssl/openssl/pull/10132#issuecomment-570664491)

LOL. You didn't expect me, did you?

richsalz commented:
Do you know the term "nerd snipe" ? :)

richsalz commented:
The sub-section issue I put into its own issue, #10809

mspncp commented:
> Do you know the term "nerd snipe" ? :)

You previously about it. But now I know why I felt so dizzy...

richsalz commented:
Whew!  Changes made
- s_client has no -certfile option, re-removed it (was in perlvars.pm)
- s_server has no -certfile option, re-removed it (was in perlvars.pm) allow_proxy_certs and issuer_checks were moved into perlvars.pm (so they shows up in several commands now). Confirming that the argument names for attime, authlevel, engine, policy, purpose, verify_{depth,email,hostname,ip,name} changed.
- ts.pod.in verifying that the argument names for CAfile,CApath,CAstore, changed. Removed -certfile (was in perlvars.pm) and restored -untrusted
- verify Similar changes to s_server; also added missing show_chain, sm2-id and sm2-hex-id, verbose, trusted, untrusted

Using ```./util/find-doc-nits -c``` to add missing options is the planned subject of a separate PR.  (I think the perl script needs some work)

richsalz commented:
Rebased to fix conflicts in s_client,s_server POD pages.

mspncp commented:
@richsalz your pr is not forgotten. I had (resp. have) some busy days, not sure when I will get around to finish the review.

mspncp commented:
Travis Errors are relevant:

```make
doc/man1/openssl-verify.pod around line 16: Unterminated B<B<...>> sequence
POD document had syntax errors at /usr/bin/pod2man5.18 line 72.
make: *** [doc/man/man1/openssl-verify.1] Error 255
```

richsalz commented:
YAF pushed.

richsalz commented:
BTW, @mspncp I am not worried about your review -- once you start, you don't let go :)

mspncp commented:
> YAF pushed.

* [Young Adult Fiction][YAF]?
* [Youth Arts Foundation][YAF]?
* [Yet Another Fotoalbum][YAF]?
* [Young Americans for Freedom][YAF]?
* [Yugoslav Air Force][YAF]?

Just too many possibilities...

[YAF]:
https://www.abbreviations.com/YAF
"NO! Yet Another Fixup!"

richsalz commented:
Context is all: Yet Another Fixup

mspncp commented:
> Context is all: Yet Another Fixup

You didn't get to see the tooltip, did you?

![image](https://user-images.githubusercontent.com/11467854/72365648-ed585580-36f8-11ea-93bd-77ab13c9e477.png)

mspncp commented:
Getting closer ;-)

The diff https://github.com/openssl/openssl/commit/c9b0315f537932764523d74d83aa04c2845997e3?diff=split shows a comparison of the generated files before and after your changes, with the options in the SYNOPSIS section sorted.

There are still some diffs in the synopsis sections.  Please take a look whether all additions/removals are desired. (I didn't try the commands). 

Two errors catched my eyes, I'll try to add them as inline comments.

richsalz commented:
I fixed the two errors.  I skimmed the diff file and it looks right to me.

My suggestion is that we merge this now and other PR's (such as #10873) will build on this.

richsalz commented:
@vdukhovni any chance you'll be able to take another (quick) look, or should I try to find someone else?

beldmit commented on doc/man1/openssl-s_server.pod.in:
```diff
@@ -692,6 +646,12 @@ by the client in binary mode.
+If the server requests a client certificate, than
+verification errors are displayed, for debugging, but the command will
+proceed unless the B<-verify_return_error> option is used.
+
```
Is it a real behavior when the client certificate is requested?

richsalz replied:
Yes, look at the verify_callback function in apps/lib/s_cb.c

beldmit commented on doc/man1/openssl.pod:
```diff
@@ -781,6 +781,257 @@ client.
+=head2 Verification Options
+
+Many OpenSSL commands verify certificates. The details of how each
+command handles errors is documented on the specific command page.
```
Nit: are documented

richsalz replied:
Nope, "each command" is singular, short for "each one of the commands" :)

kaduk replied:
But it's the *details* (plural) that are being documented here.  Er, there.  Whatever.

mspncp replied:
I agree with @beldmit and @kaduk. So that's 3:1. I suggest that I'll just fix it while merging.

beldmit commented on doc/man1/openssl.pod:
```diff
@@ -781,6 +781,257 @@ client.
+and ending in a root CA.  It is an error if the whole chain cannot be
+built up.  The chain is built up by looking up the issuers certificate
+of the current certificate.  If a certificate is found which is its own
+issuer it is assumed to be the root CA.
```
Shouldn't it be better smth like

If an issuer certificate is a certificate itself, the certificate is assumed to be the root CA.

?

richsalz replied:
I tried a slightly different wording, let me know what you think.

beldmit commented on doc/man1/openssl.pod:
```diff
@@ -781,6 +781,257 @@ client.
+considered to be valid for all purposes.
+
+The fourth, and final, step is to check the validity of the certificate
+chain. The validity period is checked against the current system time
```
Not always the current - at least can be overwritten by the `-attime` option.

richsalz replied:
I added a sentence about -attime

FdaSilvaYY commented on doc/man1/openssl-s_server.pod.in:
```diff
@@ -692,6 +646,12 @@ by the client in binary mode.
 
+{- $OpenSSL::safe::opt_v_item -}
+
+If the server requests a client certificate, than
```
then , or than ?

richsalz replied:
then. argh, fixed.

FdaSilvaYY commented on doc/man1/openssl.pod:
```diff
@@ -781,6 +781,258 @@ client.
+=item B<-ignore_critical>
+
+Normally if an unhandled critical extension is present which is not
+supported by OpenSSL the certificate is rejected (as required by RFC5280).
```
`RFC 5280 ` may be better, no ?

richsalz replied:
yes, but perhaps a separate PR (and find-doc-nits update).
```
; g grep 'RFC [0-9]' doc/man? | wc -l
     136
; g grep 'RFC[0-9]' doc/man? | wc -l
     108
```
  There are five occurrences of RFC5280 in this file so I'm going to go with the consistency.

FdaSilvaYY commented on doc/man1/openssl.pod:
```diff
@@ -781,6 +781,258 @@ client.
+
+=item B<-crl_check>
+
+Checks end entity certificate validity by attempting to look up a valid CRL.
```
this sentence is weird : 
a certificate is validated by the CRLs if its lookup in CRL has failed, or its revocation time is after check-time, no ?

What is `a valid CRL` here ? 
Did the CRL is checked against issuer of checked-certificate (mostly sure) , and/or CRL signature validity.

richsalz replied:
Yes it's not great. But again I moved it over from verify.pod  Much of this section needs copy-editing.  But this PR is 3 1/2 months old and i'm not doing anything to reset the review clock.  Consider opening an issue.  Sorry about that.

vdukhovni commented:
> @vdukhovni any chance you'll be able to take another (quick) look, or should I try to find someone else?

Cycles a bit scarce lately, I hope someone else will take you across the finish line. At least I got the ball rolling. Thanks for hanging in to the bitter end!

mattcaswell commented:
Travis red cross not relevant. See #10932

richsalz commented:
Rebased to pick up the fix in #10932; no other changes made.

mspncp commented:
Merged to master in 21d08b9ee9c0f7fabcad27b5d0b0c8c16f7dd1e9, thanks for your patience :-)

richsalz commented:
Thanks for the grammar correction, folks.  Sheesh.  3:1 you were right and i was wrong.
